File: ts/src/token_store.ts
/**
 * ROAR Protocol — Abstract token use-count store for delegation token enforcement.
 *
 * The default InMemoryTokenStore is safe for single-worker deployments.
 * For multi-worker production, use RedisTokenStore (requires ioredis package).
 *
 * Both stores are safe against race conditions within their scope:
 * - InMemoryTokenStore: JS single-threaded event loop (single process)
 * - RedisTokenStore: INCR is atomic on the Redis server (across processes)
 *
 * Mirrors python/src/roar_sdk/token_store.py exactly.
 */

// ---------------------------------------------------------------------------
// Abstract interface
// ---------------------------------------------------------------------------

export interface TokenStore {
  /**
   * Atomically increment use count.
   * Returns true if within limit, false if exhausted.
   * If maxUses is null (unlimited), always returns true and increments.
   */
  getAndIncrement(tokenId: string, maxUses: number | null): Promise<boolean>;

  /** Return current use count for a token. */
  getCount(tokenId: string): Promise<number>;
}

// ---------------------------------------------------------------------------
// In-memory store (single-process)
// ---------------------------------------------------------------------------

export class InMemoryTokenStore implements TokenStore {
  private _counts = new Map<string, number>();

  async getAndIncrement(tokenId: string, maxUses: number | null): Promise<boolean> {
    const current = this._counts.get(tokenId) ?? 0;
    if (maxUses !== null && current >= maxUses) {
      return false;
    }
    this._counts.set(tokenId, current + 1);
    return true;
  }

  async getCount(tokenId: string): Promise<number> {
    return this._counts.get(tokenId) ?? 0;
  }
}

// ---------------------------------------------------------------------------
// Redis store (multi-worker)
// ---------------------------------------------------------------------------

/**
 * Multi-worker safe token store using Redis INCR.
 *
 * Redis INCR is atomic — safe across multiple Node.js worker processes.
 * Keys are set to expire after 24 hours to prevent unbounded growth.
 *
 * Requires: npm install ioredis
 * ioredis is a dynamic import so the package is optional at module load time.
 */
export class RedisTokenStore implements TokenStore {
  private _redisUrl: string;
  private _keyPrefix: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private _client: any | null = null;

  constructor(
    redisUrl = "redis://localhost:6379/0",
    keyPrefix = "roar:tok:",
  ) {
    this._redisUrl = redisUrl;
    this._keyPrefix = keyPrefix;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private async getClient(): Promise<any> {
    if (this._client) return this._client;
    let Redis: { new(url: string): unknown };
    try {
      // Dynamic import keeps ioredis optional
      const mod = await import("ioredis");
      Redis = mod.default ?? mod;
    } catch {
      throw new Error(
        "RedisTokenStore requires the 'ioredis' package. Install: npm install ioredis",
      );
    }
    this._client = new (Redis as { new(url: string): unknown })(this._redisUrl);
    return this._client;
  }

  async getAndIncrement(tokenId: string, maxUses: number | null): Promise<boolean> {
    const redis = await this.getClient();
    const key = `${this._keyPrefix}${tokenId}`;
    // INCR is atomic; compare AFTER increment to decide if within limit
    const newCount: number = await redis.incr(key);
    // Set TTL on first use (24h default)
    if (newCount === 1) {
      await redis.expire(key, 86400);
    }
    if (maxUses !== null && newCount > maxUses) {
      return false;
    }
    return true;
  }

  async getCount(tokenId: string): Promise<number> {
    const redis = await this.getClient();
    const val: string | null = await redis.get(`${this._keyPrefix}${tokenId}`);
    return val ? parseInt(val, 10) : 0;
  }
}

File: python/src/roar_sdk/token_store.py
# -*- coding: utf-8 -*-
"""Abstract token use-count store for delegation token enforcement.

The default InMemoryTokenStore is safe for single-worker deployments.
For multi-worker production, use RedisTokenStore (requires redis package).

Both stores are safe against race conditions within their scope:
- InMemoryTokenStore: GIL-safe integer increment (single process)
- RedisTokenStore: INCR is atomic on the Redis server (across processes)
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Optional


class TokenStore(ABC):
    @abstractmethod
    def get_and_increment(self, token_id: str, max_uses: Optional[int]) -> bool:
        """Atomically increment use count. Returns True if within limit, False if exhausted.

        If max_uses is None (unlimited), always returns True and increments.
        """

    @abstractmethod
    def get_count(self, token_id: str) -> int:
        """Return current use count for a token."""


class InMemoryTokenStore(TokenStore):
    """Single-process in-memory store. NOT safe across multiple workers."""

    def __init__(self) -> None:
        self._counts: dict[str, int] = {}

    def get_and_increment(self, token_id: str, max_uses: Optional[int]) -> bool:
        current = self._counts.get(token_id, 0)
        if max_uses is not None and current >= max_uses:
            return False
        self._counts[token_id] = current + 1
        return True

    def get_count(self, token_id: str) -> int:
        return self._counts.get(token_id, 0)


class RedisTokenStore(TokenStore):
    """Multi-worker safe token store using Redis INCR.

    Redis INCR is atomic — safe across multiple uvicorn/gunicorn workers.
    Keys are set to expire after 24 hours to prevent unbounded growth.

    Use this in production deployments with more than one worker process.
    The ``redis`` package is imported lazily so the module loads fine without
    it installed — a clear ``ImportError`` is raised only when a method is
    first called.

    Requires: ``pip install roar-sdk[redis]``
    """

    def __init__(
        self,
        redis_url: str = "redis://localhost:6379/0",
        key_prefix: str = "roar:tok:",
    ) -> None:
        self._redis_url = redis_url
        self._prefix = key_prefix
        self._client = None  # lazy; created on first use

    def _get_client(self):
        if self._client is not None:
            return self._client
        try:
            import redis as _redis
        except ImportError as exc:
            raise ImportError(
                "RedisTokenStore requires the 'redis' package. "
                "Install with: pip install roar-sdk[redis]"
            ) from exc
        self._client = _redis.from_url(
            self._redis_url,
            decode_responses=True,
            socket_timeout=2.0,
            socket_connect_timeout=2.0,
        )
        return self._client

    # Lua script: atomic increment-if-below-limit with TTL.
    # Returns the new count on success, -1 if limit exceeded.
    # This prevents the race where INCR overshoots and permanently
    # blacklists a legitimate token (SEC-001).
    _LUA_INCR_IF_BELOW = """
local current = tonumber(redis.call('GET', KEYS[1]) or '0')
if ARGV[1] == 'unlimited' or current < tonumber(ARGV[1]) then
    local new_count = redis.call('INCR', KEYS[1])
    redis.call('EXPIRE', KEYS[1], 86400)
    return new_count
else
    return -1
end
"""

    def get_and_increment(self, token_id: str, max_uses: Optional[int]) -> bool:
        r = self._get_client()
        key = f"{self._prefix}{token_id}"
        limit = str(max_uses) if max_uses is not None else "unlimited"
        result = r.eval(self._LUA_INCR_IF_BELOW, 1, key, limit)
        return int(result) != -1

    def get_count(self, token_id: str) -> int:
        r = self._get_client()
        val = r.get(f"{self._prefix}{token_id}")
        return int(val) if val else 0
